### ISSUE TITLE
Decide how to keep secrets

### ISSUE THREAD

**Issue by darkdrgn2k** [tomeshnet/toronto-community-network#38]
<sup>_This initial comment is collaborative and open to modification by all._</sup>
## Task Summary

🎟️ **Re-ticketed from:** #  
📅 **Due date:** N/A  
🎯 **Success criteria:** Define a policy for secret keeping.

Who has access to what credentials, accounts, etc.
Where are they stored.

## To Do

- [x] Determine Password Management Tool
- [x] Install and Configure Password Management Tool
- [x] Provide access to Toronto Community Network contributors
- [x] Update this issue with the first draft
- [ ] Commit documentation to [docs.tomesh.net](https://github.com/tomeshnet/toronto-community-network-docs)

**Comment by infectedroot** [comment]
I've been using a tool called BitWarden for my own uses. It may have enough features for us to properly secure secrets.

Not too sure if we should also look into using something like Hashicorp Vault for any other types of secrets.

**Comment by benhylau** [comment]
I have good experience with Passbolt, making use of its teams to decide who has access to what and without central key that can see all (everyone keeps their own account key). One restriction is that we cannot store files in the vault, and in this project we may have encryption key files that need to be uploaded. KeePassX allows this but is not for teams.

@infectedroot does BitWarden satisfy the above use cases and can be self-hosted with encrypted backups?

**Comment by makeworld-the-better-one** [comment]
> One restriction is that we cannot store files in the vault, and in this project we may have encryption key files that need to be uploaded.

Is encrypting the files, then making them easy for anyone to download an option? That way the password to decrypt the file could be the secret, instead of the file itself.

**Comment by makeworld-the-better-one** [comment]
Also the [bus factor](https://en.wikipedia.org/wiki/Bus_factor) is something to keep in mind here, how to store secrets in a way that is accessible if something bad happens.

**Comment by darkdrgn2k** [comment]
FYI
I asked NYCmesh about their practices, does not keep secrets.
![image](https://user-images.githubusercontent.com/29005789/88126541-bd09dc00-cb9f-11ea-95a8-11d081f44254.png)

That is to say if some one wants a password they share it. Their passwords are scattered across people and different password managers.
![image](https://user-images.githubusercontent.com/29005789/88126565-ce52e880-cb9f-11ea-93cf-3785610ff094.png)

I suggested PassBolt to them as some of us have experience in multi-user environment with it.

Passbolt needs a Firefox or Chrome plugin to work, as far as i know does not work on mobile yet.

Two items that are in their backlog are two things we would care about.

![image](https://user-images.githubusercontent.com/29005789/88126672-13771a80-cba0-11ea-9491-f13af8c48e37.png)

**Comment by darkdrgn2k** [comment]
> 
> 
> Also the [bus factor](https://en.wikipedia.org/wiki/Bus_factor) is something to keep in mind here, how to store secrets in a way that is accessible if something bad happens.

Goes hand in hand with decentralizing knowledge from one person.

**Comment by infectedroot** [comment]
> does BitWarden satisfy the above use cases and can be self-hosted with encrypted backups?

@benhylau Ah PassBolt, I remember using that for a brief time with some people I worked with. BitWarden is very similar. There are "organizations" within BitWarden. It also cannot store files, but "secure notes" - so I paste private keys into it... just copy it once to your desktop and you're good. And yes, there's a self-hosted version of it.

> I suggested PassBolt to them as some of us have experience in multi-user environment with it.
>
> Passbolt needs a Firefox or Chrome plugin to work, as far as i know does not work on mobile yet.
> 
> Two items that are in their backlog are two things we would care about.

For accessibility - it's highly accessible. Web, Desktop, tons of browser plugins, native GUI, CLI, etc. If we wanted to store small files in the DB maybe we can base64 them into a secure note. 😝

**Comment by benhylau** [comment]
@infectedroot are you hosting it yourself? I feel we can't really support a per-user fee since the project is pretty open membership https://bitwarden.com/#organizations and we don't collect a member fee. Or if they can host the vault for free for FOSS projects. If we sort this I am happy to try it. What do others think?

**Comment by infectedroot** [comment]
@benhylau , Yes, I am hosting it for myself at home. The Community version seems to be the same as the Business version, you just specify your FQDN in the clients and you're good to go. I'm happy with either BitWarden or Passbolt, but would totally like to hear more input from others.

I find it hard to find a hosted password manager that's free for FOSS projects. Both BitWarden and Passbolt suggest to contact them for more information. It might be worth investigating so we can keep the bus factor at bay. https://github.com/tomeshnet/toronto-community-network/issues/38#issuecomment-662202540

**Comment by darkdrgn2k** [comment]
I can  spin up a VM and evaluate it.  

We will have access to a TOMESH hypervisor once  the install @ cisco is complete and can deploy production there.

**Comment by infectedroot** [comment]
If you guys want to take it for a test run... https://tomesh.infectedroot.com/

I need to add you guys to the internal organization after you sign up to share secrets between multiple people.

To use the plugins from Bitwarden, download them and then find the "Custom URL" setting and switch it to the above.

**Comment by darkdrgn2k** [comment]
@infectedroot  I'm still getting the sinning loading screen icon, cant seem to get past that :(

**Comment by darkdrgn2k** [comment]
@infectedroot maybe we should spin of a separate issue to evaluate bitwarden?

**Comment by benhylau** [comment]
I hear that @infectedroot already spun up a server for us to test? I assigned this to @infectedroot and moved the task status to `doing` to keep track I hope that's ok. What's the next step?

**Comment by infectedroot** [comment]
@darkdrgn2k Have you been able to get to the website now? It looks good on my phone, so just want other people to check access.

I encourage everyone to sign up for it, then I'll add you to the shared "organization" for us. 

Next steps might be making a policy document for it. Is there any format I need to adhere to?

**Comment by makeworld-the-better-one** [comment]
I was not able to create an account, I got this error:
> Registration not allowed or user already exists

**Comment by benhylau** [comment]
Same

<img width="318" alt="Screen Shot 2020-08-03 at 12 17 18 AM" src="https://user-images.githubusercontent.com/2261308/89133768-14ecff00-d51f-11ea-87d1-ef433db4a1fc.png">

**Comment by darkdrgn2k** [comment]
I still get the spinning loading from time to time

**Comment by infectedroot** [comment]
The spinner happens for a few seconds when loading for me... sometimes 1 to 5 seconds on pageload. I found that new signups were disabled - you guys should be able to now.

**Comment by makeworld-the-better-one** [comment]
I'm still getting the same error when I try to sign up, can anyone else confirm?

**Comment by infectedroot** [comment]
Annnnnd I didn't click save on the backend config... should be good now.

**Comment by benhylau** [comment]
@infectedroot are there offsite automated backups of the encrypted drive of passwords that can be restored if:
- the datacentre disappears
- you disappear :)

**Comment by makeworld-the-better-one** [comment]
Alright I've signed up and verified my email. Should everyone else do this, or are we just testing it out?

**Comment by benhylau** [comment]
Since this is alrdy set up, we should just use it, as long as we plan to put a backup strategy in place.

**Comment by infectedroot** [comment]
I've added @makeworld-the-better-one and @darkdrgn2k to the "Toronto Community Network" organization as admins, they can invite the rest of the group as I don't have everyone's email address. I'll see if I can start adding people from some emails I've been in.

I plan on adding the backup strategy into the policy. As an administrator of the organizations you can also export the entire vault if you guys need local copies, but I'll detail it out more in the policy.

**Comment by makeworld-the-better-one** [comment]
I got an email saying I had been invited to group, but once I clicked it, it said an admin needed to review my membership before I could actually join. So I'm still waiting on that.

**Comment by infectedroot** [comment]
Approved.

**Comment by makeworld-the-better-one** [comment]
@infectedroot https://pass.tomesh.net has been created, with a CNAME to tomesh.infectedroot.com. Could you change/fix your TLS certificate so that that subdomain connects properly?

**Comment by darkdrgn2k** [comment]
Is there any way to do a DATA EXPORT of bit warden stuff?  That way it can be pushed offsite as well.

Perhaps:
https://blog.patshead.com/2019/09/backing-up-my-bitwarden-database.html

@ASoTNetworks could you facilitate storage of an offsite backup like we did with passbolt?

**Comment by ASoTNetworks** [comment]
I can make some storage space for it

**Comment by infectedroot** [comment]
> @infectedroot https://pass.tomesh.net has been created, with a CNAME to tomesh.infectedroot.com. Could you change/fix your TLS certificate so that that subdomain connects properly?

Fixed.

**Comment by ASoTNetworks** [comment]
I have provisioned some space that can be accessed over SCP or SSHFS / SFTP. 

All I need is the SSH public key via PM on Matrix.

**Comment by makeworld-the-better-one** [comment]
@infectedroot @ASoTNetworks 

pass.tomesh.net doesn't load properly for me or @YurkoWasHere. The console reports `Loading failed for the <script> with source “https://pass.tomesh.net/app/vendor.ec1191668ddd60d16e05.js”.`. This error does not occur with tomesh.infectedroot.com

**Comment by benhylau** [comment]
Is it because it changed domains?

I tried to register but bc I have an account on the old @infectedroot test domain, it wouldn't let me, but I couldn't login with my old password either. Then I deleted the account, and was able to re-register with the same email, and it works now.

**Comment by darkdrgn2k** [comment]
Ive been having issues on both domains.

Today i was able to long twice, so maybe its fixed.

But im not part of the org i was invited to.

**Comment by benhylau** [comment]
Oh I am also not in any org, I deleted and re-registered so any previous invite would've been gone. Someone pls invite me again! We definitely need that contact info doc @darkdrgn2k mentioned in chat, so ppl know what emails to invite :)

**Comment by infectedroot** [comment]
I just re-added you to the org @benhylau . Also over the last few days I've switched the load balancer that everything was served off of... was causing those packet issues for @darkdrgn2k . 

He's also committed a few secrets too. Let me know how it works out @benhylau , then I think we can close this issue.

**Comment by benhylau** [comment]
I accepted the invite successfully but still see `There are no items to list.` under my Vault and no organization under my Organization tab.

**Comment by infectedroot** [comment]
Yeah there's a double confirm for joining a group. You should see it now.

**Comment by benhylau** [comment]
Oh weird. Perhaps a note [here](https://docs.tomesh.net/operations/secrets.html#obtaining-access) about adding to group and double confirmation?
